File: server/go/background/log_manager.go
package background

import (
	"context"
	"log/slog"
	"time"

	"github.com/luvxinc/vpn/server/store"
)

// LogManager runs nightly cleanup at 3am.
type LogManager struct {
	DB                      *store.DB
	RetentionDays           int
	MaxDomainsPerUserPerDay int
}

func NewLogManager(db *store.DB, retentionDays, maxDomains int) *LogManager {
	return &LogManager{
		DB:                      db,
		RetentionDays:           retentionDays,
		MaxDomainsPerUserPerDay: maxDomains,
	}
}

// Run sleeps until 3am then runs cleanup, forever.
func (m *LogManager) Run(ctx context.Context) {
	for {
		m.sleepUntil3AM(ctx)
		select {
		case <-ctx.Done():
			return
		default:
		}
		if err := m.RunCleanup(ctx); err != nil {
			slog.Error("log cleanup error", "err", err)
			// Retry in 1 hour on error
			select {
			case <-ctx.Done():
				return
			case <-time.After(time.Hour):
			}
		}
	}
}

// sleepUntil3AM sleeps until the next 3am, respecting context cancellation.
func (m *LogManager) sleepUntil3AM(ctx context.Context) {
	now := time.Now()
	target := time.Date(now.Year(), now.Month(), now.Day(), 3, 0, 0, 0, now.Location())
	if !target.After(now) {
		target = target.AddDate(0, 0, 1)
	}
	wait := time.Until(target)
	slog.Debug("next log cleanup", "in_hours", wait.Hours())
	select {
	case <-ctx.Done():
	case <-time.After(wait):
	}
}

// RunCleanup runs the three-step cleanup. Exported for testing.
func (m *LogManager) RunCleanup(ctx context.Context) error {
	cutoff := time.Now().AddDate(0, 0, -m.RetentionDays)

	// 1. Delete old access_log rows
	deleted, err := m.DB.DeleteOldAccessLogs(ctx, cutoff)
	if err != nil {
		return err
	}
	slog.Info("purged old access_log rows", "count", deleted, "retention_days", m.RetentionDays)

	// 2. Aggregate yesterday into traffic_daily (use UTC midnight to match DB timestamps)
	yesterday := time.Now().UTC().Truncate(24 * time.Hour).AddDate(0, 0, -1)
	if err := m.DB.UpsertTrafficDaily(ctx, yesterday); err != nil {
		return err
	}
	slog.Info("aggregated traffic_daily", "date", yesterday.Format("2006-01-02"))

	// 3. Cap access_log per user per day
	if err := m.DB.CapAccessLog(ctx, cutoff, m.MaxDomainsPerUserPerDay); err != nil {
		return err
	}
	slog.Info("log cleanup complete")
	return nil
}

File: server/services/log_manager.py
"""Background task: nightly log cleanup and traffic_daily aggregation."""
import asyncio
import logging
from datetime import date, datetime, timedelta, timezone

log = logging.getLogger("weiai.log_manager")


async def cleanup_loop(app):
    """Runs forever. At 3am each night, purge old logs and aggregate daily stats."""
    while True:
        try:
            await _sleep_until_3am()
            await _run_cleanup(app)
        except asyncio.CancelledError:
            return
        except Exception as e:
            log.error(f"Log cleanup error: {e}")
            await asyncio.sleep(3600)  # retry in 1 hour on error


async def _sleep_until_3am():
    now = datetime.now()
    target = now.replace(hour=3, minute=0, second=0, microsecond=0)
    if target <= now:
        target += timedelta(days=1)
    wait_seconds = (target - now).total_seconds()
    log.debug(f"Next log cleanup in {wait_seconds/3600:.1f} hours")
    await asyncio.sleep(wait_seconds)


async def _run_cleanup(app):
    import config_loader
    cfg = config_loader.get()
    db = app.state.db
    retention_days = cfg["log"]["retention_days"]
    max_domains = cfg["log"]["max_domains_per_user_per_day"]

    # 1. Delete access_log rows older than retention_days
    cutoff = datetime.now() - timedelta(days=retention_days)
    deleted = await db.fetchval(
        "DELETE FROM access_log WHERE access_hour < $1 RETURNING COUNT(*)",
        cutoff,
    )
    log.info(f"Purged {deleted or 0} access_log rows older than {retention_days} days")

    # 2. Aggregate yesterday's sessions into traffic_daily
    yesterday = date.today() - timedelta(days=1)
    await db.execute(
        """INSERT INTO traffic_daily (user_id, date, upload_bytes, download_bytes)
           SELECT user_id,
                  $1::date,
                  COALESCE(SUM(upload_bytes), 0),
                  COALESCE(SUM(download_bytes), 0)
           FROM sessions
           WHERE DATE(connected_at) = $1
           GROUP BY user_id
           ON CONFLICT (user_id, date)
           DO UPDATE SET
             upload_bytes   = EXCLUDED.upload_bytes,
             download_bytes = EXCLUDED.download_bytes
        """,
        yesterday,
    )
    log.info(f"Aggregated traffic_daily for {yesterday}")

    # 3. Cap access_log per user per day to max_domains (keep highest-traffic hosts)
    await db.execute(
        """DELETE FROM access_log
           WHERE id IN (
             SELECT id FROM (
               SELECT id,
                      ROW_NUMBER() OVER (
                        PARTITION BY user_id, DATE(access_hour)
                        ORDER BY download_bytes DESC, upload_bytes DESC
                      ) AS rn
               FROM access_log
               WHERE access_hour >= $1
             ) ranked
             WHERE rn > $2
           )
        """,
        cutoff,
        max_domains,
    )
    log.info("Log cleanup complete")
